src/ScheduleFunctions.cpp

```diff
@@ -1099,9 +1113,45 @@ class InjectFunctionRealization : public IRMutator {
     const LoopLevel &store_level;
 
     Stmt build_realize(Stmt s, const Function &func, bool is_output) {
+        const string &name = func.name();
+        if (_funcs_need_alloc_mutex.find(func) != _funcs_need_alloc_mutex.end()) {
+            // Place an Allocate node here and a call to initialization to create a mutex buffer
+            debug(1) << "Injecting allocation of mutex buffer for " << name << '\n';
```

Comment:
I think this is way too early to be injecting the mutex buffers. Can they be introduced after storage flattening in a dedicated pass? Then you won't have to inject dependencies on symbols that don't exist yet.

Reply:
Yes they can. This is done.